src/rgw/rgw_rest_s3.h

```diff
@@ -655,30 +655,62 @@ static inline int valid_s3_bucket_name(const string& name, bool relaxed=false)
   // This function enforces Amazon's spec for bucket names.
   // (The requirements, not the recommendations.)
   int len = name.size();
+  int max = (relaxed ? 255 : 63);
+
   if (len < 3) {
     // Name too short
     return -ERR_INVALID_BUCKET_NAME;
-  } else if (len > 255) {
+  } else if (len > max) {
     // Name too long
     return -ERR_INVALID_BUCKET_NAME;
   }
 
-  // bucket names must start with a number, letter, or underscore
+  // bucket names must start with a number or letter
   if (!(isalpha(name[0]) || isdigit(name[0]))) {
     if (!relaxed)
       return -ERR_INVALID_BUCKET_NAME;
+      // do we have to support underscore? Maybe needed for backward
+      // compatibility.
     else if (!(name[0] == '_' || name[0] == '.' || name[0] == '-'))
       return -ERR_INVALID_BUCKET_NAME;
   }
 
+  // bucket names cannot end with dash
+  if (name[len-1] == '-')
+    if (!relaxed)
+      return -ERR_INVALID_BUCKET_NAME;
+
   for (const char *s = name.c_str(); *s; ++s) {
     char c = *s;
-    if (isdigit(c) || (c == '.'))
-      continue;
-    if (isalpha(c))
+    if (isdigit(c))
       continue;
-    if ((c == '-') || (c == '_'))
+
+    if (isalpha(c)) {
+      // name cannot contain uppercase letters
+      if (relaxed || islower(c))
+     	continue;
+    }
+
+    if (c == '_')
+      // name cannot contain underscore
+      if (relaxed)
+	continue;
+
+    if (c == '-')
       continue;
+   
+    if (c == '.') {
+      // name cannot have consecutive periods or dashes
+      // adjacent to periods
+      // ensure s is neither the first nor the last character
+      if (!relaxed && s && *s) {
+	char p = *(s-1);
+	char n = *(s+1);
+	if ((p != '-') && (n != '.') && (n != '-'))
+	  continue;
+      }
+    }
```

Comment:
when `relaxed == true`, this rejects any bucket name with a period in it (so is _more_ strict, rather than less)

and when `!relaxed` and given a `name = ".foo"`, it will still dereference the memory at `name[-1]`

Reply:
> when `relaxed == true`, this rejects any bucket name with a period in it (so is _more_ strict, rather than less)

Done.

> 
> and when `!relaxed` and given a `name = ".foo"`, it will still dereference the memory at `name[-1]`

There are two places which I think will avoid this issue.

- At line660
 669   // bucket names must start with a number or letter
 670   if (!(isalpha(name[0]) || isdigit(name[0]))) {
 671     if (!relaxed)
 672       return -ERR_INVALID_BUCKET_NAME;

 - At line706:
       if (!relaxed && s && *s) {
We check if 's' is not '0' (first character) and *s (for last character).

Please let me know if I missed anything else here.